firedrake/solving_utils.py

```diff
@@ -106,19 +107,53 @@ def __init__(self, problem, mat_type, pmat_type, appctx=None,
         self.F = problem.F
         self.J = problem.J
 
-        if mat_type != pmat_type or problem.Jp is not None:
+        # For Jp to equal J, bc.Jp must equal bc.J for all EquationBC objects.
+        Jp_eq_J = all(b.Jp_eq_J for b in problem)
```

Comment:
Is this right? Do you mean `for b in problem.bcs`?

Reply:
So, as I mentioned eariler, I think NLVP class and EquationBC class have a lot in common by nature, and I thought it makes sense to add __iter__ method in the same way in NLVP class.

Reply:
While I agree that they are quite similar, iterating over the `EquationBC` gives back a bunch of things that are "like BCs". Whereas iterating over a `NLVP` gives back an `NLVP` and then a bunch of things "like BCs". Which feels less natural to me.

Reply:
So, my idea is to make a parent class for those two, say `Equation` class, which contains, e.g. `F`, `J`, `Jp`, `bc`, and `__iter__` method (but not in this PR at least).  The motivation for this is that even a 3D PDE can be regarded as an `EquationBC` of a fictitious 4D problem, which I believe simplifies the code by removing the need to separately deal with the 'domain' and the 'boundary.'  For instance, here, I can check `Jp_eq_J` in a monolithic way, which I think is more natural.  
`NLVP` currently has more things, so I would like to discuss more if we can really do this, but this was basically what I was thinking.